src/main/java/edu/cornell/kfs/fp/service/impl/RecurringDisbursementVoucherDocumentServiceImpl.java

```diff
@@ -311,7 +339,72 @@ private RecurringDisbursementVoucherPDPStatus buildRecurringDisbursementVoucherP
         }
         return paymentDetails;
     }
- 
+    
+    private void noteChangeOnRecurringDV(RecurringDisbursementVoucherDocument recurringDV, String noteText, Set<String> setOfStrings) {
+        if (setOfStrings != null && !setOfStrings.isEmpty()) {
+            Note paymentGroupNote = buildNoteBase();
+            paymentGroupNote.setNoteText(noteText + StringUtils.join(setOfStrings, ", "));;
+            recurringDV.addNote(paymentGroupNote);
+            getBusinessObjectService().save(recurringDV);
+        }
+    }
+    
+    private Note buildNoteBase() {
+        Note noteBase = new Note();
+        noteBase.setAuthorUniversalIdentifier(GlobalVariables.getUserSession().getPerson().getPrincipalId());
+        noteBase.setNotePostedTimestampToCurrent();
+        return noteBase;
+    }
+    
+    @Override
+    public boolean isPaymentCancelable(Person user, String paymentDetailStatus) {
+        return paymentDetailStatus.equalsIgnoreCase(PdpConstants.PaymentStatusCodes.OPEN) 
+                && getRecurringDisbursementVoucherPaymentMaintenanceService().hasCancelPermission(user);
+    }
+    
+    @Override
+    public Set<String> cancelDisbursementVouchers(RecurringDisbursementVoucherDocument recurringDisbursementVoucherDocument, final String cancelMessage) {
+        Set<String> canceledDVs = new HashSet<String>();
+        for (RecurringDisbursementVoucherDetail detail : recurringDisbursementVoucherDocument.getRecurringDisbursementVoucherDetails()) {
+            CuDisbursementVoucherDocument dv = retrieveCUDisbursementVoucherDocument(detail);
```

Comment:
This code has the potential of retrieving the same document twice: Once to check if it's cancelable and again to perform the actual cancellation. Because the isDVCancelable() check only appears to require a permission check and a route status check, you could instead use a method such as RouteHeaderService.getDocumentStatus() to get the status for the can-cancel check, and then wait until the actual cancellation point to retrieve the full document.